presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPlanWithEmptyInput.java

```diff
@@ -197,6 +205,51 @@ else if (isEmptyNode(rewrittenRight)) {
             return node.replaceChildren(ImmutableList.of(rewrittenLeft, rewrittenRight));
         }
 
+        @Override
+        public PlanNode visitSequence(SequenceNode node, RewriteContext<Void> context)
+        {
+            List<PlanNode> cteProducers = node.getCteProducers();
+            List<PlanNode> newSequenceChildrenList = new ArrayList<>();
+            // Visit in the order of execution
+            for (int i = cteProducers.size() - 1; i >= 0; i--) {
+                PlanNode rewrittenProducer = context.rewrite(cteProducers.get(i));
+                if (!isEmptyNode(rewrittenProducer)) {
+                    newSequenceChildrenList.add(rewrittenProducer);
+                }
+            }
+            PlanNode rewrittenPrimarySource = context.rewrite(node.getPrimarySource());
+            if (isEmptyNode(rewrittenPrimarySource) || newSequenceChildrenList.isEmpty()) {
+                return rewrittenPrimarySource;
+            }
+            // Reverse order for execution
+            Collections.reverse(newSequenceChildrenList);
+            // Add the primary source at the end of the list
+            newSequenceChildrenList.add(rewrittenPrimarySource);
+            return node.replaceChildren(newSequenceChildrenList);
+        }
+
+        @Override
+        public PlanNode visitCteProducer(CteProducerNode node, RewriteContext<Void> context)
+        {
+            PlanNode rewrittenSource = context.rewrite(node.getSource());
+            if (isEmptyNode(rewrittenSource)) {
+                // Remove CTE materialization from session
+                // This will be used to convert consumer to values and in further optimizations
+                session.getCteInformationCollector().disallowCteMaterialization(node.getCteId());
+                return convertToEmptyValuesNode(node);
+            }
+            return node.replaceChildren(ImmutableList.of(rewrittenSource));
+        }
+
+        @Override
+        public PlanNode visitCteConsumer(CteConsumerNode node, RewriteContext<Void> context)
+        {
+            if (!session.getCteInformationCollector().getCteInformationMap().get(node.getCteId()).isMaterialized()) {
```

Comment:
What happens when the cte is set to not materialize due to HEURISTIC strategy? Can you add such a test case?

Reply:
For heuristic strategy the logicalCteOptimizer will make sure that their producers and consumers will never be present in the plan tree.

Reply:
So we don't need the test case for that here since we already have such removal tests [here](https://github.com/prestodb/presto/blob/master/presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestLogicalCteOptimizer.java#L341). This testing is about materialized ctes and how this optimizer handles them